pkg/plugin/stream_reader.go

```diff
@@ -0,0 +1,75 @@
+/*
+Copyright 2017 Heptio Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package plugin
+
+import (
+	"bytes"
+	"io"
+)
+
+// ReceiveFunc is a function that either returns a slice
+// of an arbitrary number of bytes OR an error. Returning
+// an io.EOF means there is no more data to be read; any
+// other error is considered an actual error.
+type ReceiveFunc func() ([]byte, error)
+
+// CloseFunc is used to signal to the source of data that
+// the StreamReadCloser has been closed.
+type CloseFunc func() error
+
+// StreamReadCloser wraps a ReceiveFunc and a CloseSendFunc
+// to implement io.ReadCloser.
+type StreamReadCloser struct {
+	buf     *bytes.Buffer
+	receive ReceiveFunc
+	close   CloseFunc
+}
+
+func (s *StreamReadCloser) Read(p []byte) (n int, err error) {
+	for {
+		// if buf exists and holds at least as much as we're trying to read,
+		// read from the buffer
+		if s.buf != nil && s.buf.Len() >= len(p) {
```

Comment:
I still think you want to check `cap`, not `len`. If `p` is a slice that has len=0, cap=100, and `s.buf.Len()` is 50, this doesn't match your comment. We're trying to read up to 100 (cap), and you only want to go down this path if `s.buf.Len()` is >= 100, right?

Reply:
Also, if we have data in the buffer, does it make sense to flush it to `p` asap, or always to try to fill the buffer up to `cap(p)` (assuming you agree with ^^^^) before flushing it?

Reply:
So the godoc for io.Reader says:

```
Read reads up to len(p) bytes into p. It returns the number of bytes read (0 <= n <= len(p))...
```

Reply:
Ok :-)